<a href="https://colab.research.google.com/github/nimu77/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module3-join-and-reshape-data/LS_DS_113_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [1]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-04-08 19:19:16--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.177.77
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.177.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  45.0MB/s    in 4.6s    

2020-04-08 19:19:21 (42.7 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [2]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [3]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [4]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [6]:
##### YOUR CODE HERE #####
import pandas as pd
aisle = pd.read_csv('aisles.csv')
print(aisle.shape)
aisle.head()

(134, 2)


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [7]:
departments = pd.read_csv('departments.csv')
print(departments.shape)
departments.head()

(21, 2)


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [8]:
order_products_prior = pd.read_csv('order_products__prior.csv')
print(order_products_prior.shape)
order_products_prior.head()

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [9]:
order_products_train = pd.read_csv('order_products__train.csv')
print(order_products_train.shape)
order_products_train.head()

(1384617, 4)


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [10]:
orders = pd.read_csv('orders.csv')
print(orders.shape)
orders.head()

(3421083, 7)


order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [11]:
products = pd.read_csv('products.csv')
print(products.shape)
products.head()

(49688, 4)


product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [14]:
products['product_name'].value_counts(ascending=True)

Rtd 3 2 1 Rich Choc Royal                                1
The Complete Cookie Oatmeal Raisin                       1
Shredded Wheat                                           1
Butterscotch Chips Baking Pieces                         1
For Spit-Up with Iron Infant Formula                     1
                                                        ..
Low-Fat Mango on the Bottom Greek Yogurt                 1
Korean Style Barbecue Bulgogl Sauce                      1
Organic Grass-Fed Real Aged Cheddar & Macaroni Shells    1
Toothpaste, Ayurvedic Herbal, Mint-Free                  1
Sesame Roasted Seaweed Snacks                            1
Name: product_name, Length: 49688, dtype: int64

In [18]:
orders_products = pd.concat([order_products_train[['order_id', 'product_id', 'add_to_cart_order']],
                            order_products_prior[['order_id', 'product_id', 'add_to_cart_order']]])
orders_products.head()

order_id  product_id  add_to_cart_order
0         1       49302                  1
1         1       11109                  2
2         1       10246                  3
3         1       49683                  4
4         1       43633                  5

In [0]:
products = products[['product_id', 'product_name']]

In [21]:
products.head()

product_id                                       product_name
0           1                         Chocolate Sandwich Cookies
1           2                                   All-Seasons Salt
2           3               Robust Golden Unsweetened Oolong Tea
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...
4           5                          Green Chile Anytime Sauce

In [23]:
merged = pd.merge(orders_products, products, on='product_id')
merged.head()

order_id  product_id  add_to_cart_order      product_name
0         1       49302                  1  Bulgarian Yogurt
1    816049       49302                  7  Bulgarian Yogurt
2   1242203       49302                  1  Bulgarian Yogurt
3   1383349       49302                 11  Bulgarian Yogurt
4   1787378       49302                  8  Bulgarian Yogurt

In [33]:
merged.shape

(33819106, 4)

In [31]:
merged.tail()

order_id  ...                    product_name
33819101   3265099  ...        Gourmet Burger Seasoning
33819102   3361945  ...        Gourmet Burger Seasoning
33819103   3267201  ...  Piquillo & Jalapeno Bruschetta
33819104   3393151  ...                  Original Jerky
33819105   3400803  ...     Flatbread Pizza All Natural

[5 rows x 4 columns]

In [118]:
merged.groupby('product_name').size().sort_values(ascending=False).head(10)

product_name
Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
dtype: int64

### Top 10 ordered products:
1. Banana : 491291
2. Bag of Organic Bananas: 394930
3. Organic Strawberries: 275577
4. Organic Baby Spinach: 251705
5. Organic Hass Avocado: 220877
6. Organic Avocado: 184224
7. Large Lemon: 160792
8. Strawberries: 149445
9. Limes: 146660
10. Organic Whole Milk: 142813

## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [40]:
import pandas as pd
import numpy as np

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [41]:
# table2 -> tidy
table2 = table2.reset_index()
table2

index  John Smith  Jane Doe  Mary Johnson
0  treatmenta         NaN      16.0           3.0
1  treatmentb         2.0      11.0           1.0

In [42]:
tidy = table2.melt(id_vars='index', value_vars=['John Smith', 'Jane Doe', 'Mary Johnson'])
tidy

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [43]:
tidy.columns = ['trt', 'name', 'result']
tidy

trt          name  result
0  treatmenta    John Smith     NaN
1  treatmentb    John Smith     2.0
2  treatmenta      Jane Doe    16.0
3  treatmentb      Jane Doe    11.0
4  treatmenta  Mary Johnson     3.0
5  treatmentb  Mary Johnson     1.0

In [44]:
tidy['trt'] = tidy['trt'].str.replace('treatment', '')
tidy

trt          name  result
0   a    John Smith     NaN
1   b    John Smith     2.0
2   a      Jane Doe    16.0
3   b      Jane Doe    11.0
4   a  Mary Johnson     3.0
5   b  Mary Johnson     1.0

In [51]:
# tidy -> table2
wide = tidy.pivot_table(index='trt', columns='name', values='result')
wide

name  Jane Doe  John Smith  Mary Johnson
trt                                     
a         16.0         NaN           3.0
b         11.0         2.0           1.0

In [0]:
import seaborn as sns

In [49]:
flights = sns.load_dataset('flights')
flights

year      month  passengers
0    1949    January         112
1    1949   February         118
2    1949      March         132
3    1949      April         129
4    1949        May         121
..    ...        ...         ...
139  1960     August         606
140  1960  September         508
141  1960    October         461
142  1960   November         390
143  1960   December         432

[144 rows x 3 columns]

In [52]:
# Flights Pivot Table
wide = flights.pivot_table(index='year', columns='month', values='passengers')
wide

month  January  February  March  April  ...  September  October  November  December
year                                    ...                                        
1949       112       118    132    129  ...        136      119       104       118
1950       115       126    141    135  ...        158      133       114       140
1951       145       150    178    163  ...        184      162       146       166
1952       171       180    193    181  ...        209      191       172       194
1953       196       196    236    235  ...        237      211       180       201
1954       204       188    235    227  ...        259      229       203       229
1955       242       233    267    269  ...        312      274       237       278
1956       284       277    317    313  ...        355      306       271       306
1957       315       301    356    348  ...        404      347       305       336
1958       340       318    362    348  ...        404      359       310       337
1959       360       342    406    396  ...        463      407       362       405
1960       417       391    419    461  ...        508      461       390       432

[12 rows x 12 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [53]:
##### YOUR CODE HERE #####
products.head()

product_id                                       product_name
0           1                         Chocolate Sandwich Cookies
1           2                                   All-Seasons Salt
2           3               Robust Golden Unsweetened Oolong Tea
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...
4           5                          Green Chile Anytime Sauce

In [54]:
merged.head()

order_id  product_id  add_to_cart_order      product_name
0         1       49302                  1  Bulgarian Yogurt
1    816049       49302                  7  Bulgarian Yogurt
2   1242203       49302                  1  Bulgarian Yogurt
3   1383349       49302                 11  Bulgarian Yogurt
4   1787378       49302                  8  Bulgarian Yogurt

In [95]:
orders.shape

(3421083, 7)

In [55]:
orders.head()

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [94]:
orders.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [57]:
merged_again = pd.merge(merged, orders[['order_id','order_hour_of_day']], on='order_id')
merged_again.head()

order_id  ...  order_hour_of_day
0         1  ...                 10
1         1  ...                 10
2         1  ...                 10
3         1  ...                 10
4         1  ...                 10

[5 rows x 5 columns]

In [59]:
merged_again.shape

(33819106, 5)

In [102]:
grouped_df = orders.groupby(["order_hour_of_day", 'order_id'])["order_number"].aggregate("count").reset_index()
grouped_df['order_number'].sum()

3421083

In [103]:
new_grouped_df = grouped_df[grouped_df['order_hour_of_day'] > 12]
new_grouped_df

order_hour_of_day  order_id  order_number
1467662                 13        28             1
1467663                 13        35             1
1467664                 13        50             1
1467665                 13        63             1
1467666                 13        71             1
...                    ...       ...           ...
3421078                 23   3420918             1
3421079                 23   3420933             1
3421080                 23   3420973             1
3421081                 23   3421004             1
3421082                 23   3421020             1

[1953421 rows x 3 columns]

In [104]:
merged.head()

order_id  product_id  add_to_cart_order      product_name
0         1       49302                  1  Bulgarian Yogurt
1    816049       49302                  7  Bulgarian Yogurt
2   1242203       49302                  1  Bulgarian Yogurt
3   1383349       49302                 11  Bulgarian Yogurt
4   1787378       49302                  8  Bulgarian Yogurt

In [0]:
merge = pd.merge(new_grouped_df, merged, on='order_id', how='left')

In [106]:
merge

order_hour_of_day  ...                                       product_name
0                        13  ...                               Organic Hass Avocado
1                        13  ...                                        Large Lemon
2                        13  ...                       Organic Heavy Whipping Cream
3                        13  ...                         Whole Organic Omega 3 Milk
4                        13  ...                        Cultured Low Fat Buttermilk
...                     ...  ...                                                ...
19210804                 23  ...                                 Pure Coconut Water
19210805                 23  ...                        Organic Reduced Fat 2% Milk
19210806                 23  ...  Organic Salted Individually Wrapped Quarters B...
19210807                 23  ...                   Organic Cocoa Crunch Protein Bar
19210808                 23  ...                                Organic Chili Spicy

[19210809 rows x 6 columns]

In [114]:
merge.groupby('order_hour_of_day')['product_name'].size().reset_index()

order_hour_of_day  product_name
0                  13       2784168
1                  14       2817301
2                  15       2787163
3                  16       2653804
4                  17       2191771
5                  18       1718801
6                  19       1320804
7                  20       1020163
8                  21        832806
9                  22        663539
10                 23        420489

In [62]:
merged_again.groupby('product_name')['order_hour_of_day'].size().sort_values(ascending=False)

product_name
Banana                                         491291
Bag of Organic Bananas                         394930
Organic Strawberries                           275577
Organic Baby Spinach                           251705
Organic Hass Avocado                           220877
                                                ...  
Organic Ice Cream Birthday Cake                     1
11.3 Oz. Oreo Fudge Creme Double Chocolate          1
Cld/Flu Van Chrry                                   1
12 Inch Taper Candle White                          1
All Natural Stevia Liquid Extract Sweetener         1
Name: order_hour_of_day, Length: 49685, dtype: int64

In [68]:
merged_again.groupby('order_hour_of_day')['product_name'].size()

order_hour_of_day
0      228031
1      121412
2       72660
3       53759
4       55714
5       91909
6      302642
7      928239
8     1787359
9     2550569
10    2874905
11    2852701
12    2732599
13    2778054
14    2810918
15    2780731
16    2647695
17    2186409
18    1714445
19    1317576
20    1017958
21     831183
22     662053
23     419585
Name: product_name, dtype: int64

In [70]:
pd.crosstab(merged_again['order_hour_of_day'] > 12, merged_again['product_name'])

product_name       #2 Coffee Filters  ...  with a Splash of Pineapple Coconut Water
order_hour_of_day                     ...                                          
0                                  3  ...                                         0
1                                  4  ...                                         0
2                                  4  ...                                         0
3                                  1  ...                                         0
4                                  0  ...                                         0
5                                  1  ...                                         0
6                                  6  ...                                         0
7                                 28  ...                                         0
8                                 42  ...                                         4
9                                 59  ...                                         1
10                                65  ...                                         7
11                                71  ...                                         5
12                                62  ...                                         7
13                                69  ...                                         6
14                                64  ...                                         7
15                                75  ...                                         5
16                                65  ...                                         5
17                                55  ...                                         4
18                                34  ...                                         3
19                                29  ...                                         1
20                                26  ...                                         1
21                                19  ...                                         1
22                                 9  ...                                         0
23                                 8  ...                                         0

[24 rows x 49685 columns]

In [0]:
new_df = pd.crosstab(merged_again['order_hour_of_day'] > 12, merged_again['product_name'])

In [80]:
new_df.head()

product_name       #2 Coffee Filters  ...  with a Splash of Pineapple Coconut Water
order_hour_of_day                     ...                                          
False                            346  ...                                        24
True                             453  ...                                        33

[2 rows x 49685 columns]

In [82]:
new_df.transpose()

order_hour_of_day                                   False  True 
product_name                                                    
#2 Coffee Filters                                     346    453
#2 Cone White Coffee Filters                            8      6
#2 Mechanical Pencils                                   6      3
#4 Natural Brown Coffee Filters                       138    109
& Go! Hazelnut Spread + Pretzel Sticks                 81     93
...                                                   ...    ...
with Xylitol Unwrapped Original Flavor 50 Stick...     19     22
with Xylitol Unwrapped Spearmint 50 Sticks Suga...     17     40
with Xylitol Watermelon Twist 18 Sticks Sugar F...     35     39
with a Splash of Mango Coconut Water                   15     12
with a Splash of Pineapple Coconut Water               24     33

[49685 rows x 2 columns]

In [90]:
new_df.columns = ['Before noon', 'Afternoon']

ValueError: ignored

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####